### PR TITLE
CoW: Add warning for replace with inplace

### PR DESCRIPTION
**Pull request by phofl**
xref https://github.com/pandas-dev/pandas/issues/56019

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -7773,6 +7774,17 @@ def replace(
                         ChainedAssignmentError,
                         stacklevel=2,
                     )
+            elif not PYPY and not using_copy_on_write():
+                ctr = sys.getrefcount(self)
+                ref_count = REF_COUNT
+                if isinstance(self, ABCSeries) and hasattr(self, "_cacher"):
+                    ref_count += 1
```
Comment (jorisvandenbossche) [inline pandas/core/generic.py:7781]: The problem might be that in the warning mode, this is not the case? (so that might need to add a `not warn_copy_on_write()` to this if block updating the ref_count

Although the tests are not failing ..

File: pandas/tests/copy_view/test_replace.py [hunk 2]
```diff
@@ -395,6 +396,17 @@ def test_replace_chained_assignment(using_copy_on_write):
         with tm.raises_chained_assignment_error():
             df[["a"]].replace(1, 100, inplace=True)
         tm.assert_frame_equal(df, df_orig)
+    else:
+        with tm.assert_produces_warning(FutureWarning, match="inplace method"):
+            with option_context("mode.chained_assignment", None):
```
Comment (jorisvandenbossche) [inline pandas/tests/copy_view/test_replace.py:401]: Does it otherwise also raise a SettingWithCopyWarning?

Comment (phofl) [reply pandas/tests/copy_view/test_replace.py:401]: yes, [["a"]] currently copies

File: pandas/core/generic.py (hunk 1 above)
Comment (phofl) [reply pandas/core/generic.py:7781]: I don't understand your comment, why would this not be the case?

I double checked and it seems to work in warning mode

Comment (jorisvandenbossche) [reply pandas/core/generic.py:7781]: Related to https://github.com/pandas-dev/pandas/pull/55838#discussion_r1384697184, I thought I had disabled the item cache for the warning mode, which I would think to affect the count.

Comment (phofl) [reply pandas/core/generic.py:7781]: I double checked this locally, the cache is still populated

Comment (jorisvandenbossche) [reply pandas/core/generic.py:7781]: Ah, no, the `_cacher` points to the DataFrame (in a case like `s = df["col"]`), so it doesn't increase the ref count for `s` (which is what is relevant for chained setitem detection). And also it uses a weakref, so shouldn't actually increase the ref count?

The `hasattr(self, "_cacher")` is kind of a check whether the Series is derived from a DataFrame? 
But I _think_ that the reason we need to increase the ref count is still because of `_item_cache`. So maybe a more "correct" check (although they will give the same result) would be:

```suggestion
                if isinstance(self, ABCSeries) and not (using_copy_on_write() or warn_copy_on_write()):
                    ref_count += 1
```

Comment (jorisvandenbossche) [reply pandas/core/generic.py:7781]: Hmm, that suggestion gives a bunch of failures (false positives) in `pandas/tests/series/methods/test_replace.py`

Comment (jorisvandenbossche) [reply pandas/core/generic.py:7781]: Which is logical, because that's the whole point with `_item_cache`: you don't know when it is populated or when not in the non-CoW case. So you can't use a single fixed REF_COUNT value to check, this actually depends on the circumstances. 

And so what you have (checking `_cacher`) is a way to check if the Series object is derived from a DataFrame with simple indexing. So that's probably fine? Or are there ways to get a Series that does _not_ populate the cacher?

Comment (jorisvandenbossche) [reply pandas/core/generic.py:7781]: > Or are there ways to get a Series that does _not_ populate the cacher?

Getting a single column with `["col"]`, `.loc[:, "col"]`, `.iloc[:, 0]`, `.get("col")`, `.xs("col", axis=1)` all populate the cacher. 

In theory you can get a Series as a result of a calculation that doesn't do this (`df.mean().replace(..)`, but of course that never could have done something useful so we don't need to care about that. We only need to care about getting a Series that is _view_. 

So to summarize: your fix is probably fully correct, I only didn't understand why ;) You might want to add a comment about it.

File: pandas/core/generic.py [hunk 3]
```diff
@@ -7773,6 +7774,17 @@ def replace(
                         ChainedAssignmentError,
                         stacklevel=2,
                     )
+            elif not PYPY and not using_copy_on_write():
+                ctr = sys.getrefcount(self)
+                ref_count = REF_COUNT
+                if isinstance(self, ABCSeries) and hasattr(self, "_cacher"):
```
[the remark this answers is not in the document]
Comment (phofl) [reply pandas/core/generic.py:7780]: I am fine with adding this here. Maybe adding a link to this comment for the other prs?

File: pandas/core/generic.py (hunk 1 above)
Comment (phofl) [reply pandas/core/generic.py]: Yes your conclusion is correct as far as I can tell. That was the reason why I added the hasattr check. Sorry for omitting this information. 

Added a comment

**Comment by phofl** [comment]
Are you ok with merging this so that I can adjust the others?

**Comment by jorisvandenbossche** [comment]
Yes, merge away

Op ma 20 nov. 2023 22:32 schreef Patrick Hoefler ***@***.***>:
